ssz_serialization/types.nim

```diff
@@ -307,28 +333,32 @@ func clearCache*(a: var HashList) =
   # contents
   for c in a.hashes.mitems(): clearCache(c)
 
-func growHashes*(a: var HashList) =
-  ## Ensure that the hash cache is big enough for the data in the list - must
-  ## be called whenever `data` grows.
+func resizeHashes*(a: var HashList) =
+  ## Ensure the hash cache is the correct size for the data in the list - must
+  ## be called whenever `data` shrinks or grows.
   let
     leaves = int(
       chunkIdx(a, a.data.len() + dataPerChunk(a.T) - 1))
     newSize = 1 + cacheNodes(a.maxDepth, leaves)
 
-  if a.hashes.len >= newSize:
+  # Growing might be because of add(), addDefault(), or in-place reading of a
+  # larger HashList. In-place reading of a smaller HashList causes shrinking.
+  if a.hashes.len == newSize:
     return
 
   var
-    newHashes = newSeq[Digest](newSize)
+    newHashes = newSeqWith(newSize, uninitSentinel)
```

Comment:
This is a really unfortunate (but necessary) change. In principle, it should be neutral -- instead of 0-init, something-else-init. But Nim does the zero-init regardless.

https://github.com/status-im/nim-ssz-serialization/blob/04fd15bfc1e559eb307600f1acccb5876316090d/ssz_serialization/types.nim#L181-L191

also shows this problem.

Reply:
https://github.com/nim-lang/Nim/issues/22554

Reply:
wouldn't an all-zeroes sentinel work?

Reply:
There are two kinds of "uncached" nodes in this tree.

- ones where a node being cleared implies the parent was cleared; and
- ones where a node being cleared does not imply the parent was cleared.

The former allow for this tree-walking optimization, while the latter do not. `clearCaches` creates the former, and this is the main goal of the `ssz_serialization/types` changes: allow per-index invalidating to be reasonably efficient regardless of Merkle tree height.

`resizeHashes`/`growHashes`, when growing a cached hashes list, creates the latter, so the tree-walking optimization is not correct in those cases.

This does mean that, actually, the current code is slightly wrong in the case where the last 24 bytes of the digest happens to line up with the last 24 bytes of the sentinel before `clearCaches` gets it.

But this still means that because
https://github.com/status-im/nim-ssz-serialization/blob/04fd15bfc1e559eb307600f1acccb5876316090d/ssz_serialization/types.nim#L243-L244
probably then has to write some additional bytes, `resizeHashes` itself can just use zero-init, and the additional cost moves from `resizeHashes` to `clearCache`.

My understanding is that typically, memory systems work in (typically 64-byte) cacheline granularities, so writing, say, 16 bytes shouldn't be particularly more costly than writing 8 bytes, but remains to be benchmarked.

Reply:
Exploring an approach which kept an all-zeros sentinel, the problem is that it has to be distinguished somehow from the situation where it is okay to stop walking the tree. The two relevant places for this are `clearCache` (at least, the `HashList` version in this PR) and `resizeCache`. Currently the API treates `resizeCache` as more of an internal API than `clearCache`/`clearCaches`.

One issue with using `clearCache`/`clearCaches` for this purpose is code patterns such as
https://github.com/status-im/nim-ssz-serialization/blob/cf2319bdff5c63d04b320f830dd7ea8b7e152bc6/ssz_serialization/types.nim#L383-L393

which `resize` then `clear`, but do not call `hash_tree_root` to then fill in those cleared nodes. The `clear` then (because it has to have some distinguishable pattern) leaves behind a marker which is distinguishable in terms of strict `hashList.hashes` equality even though the Merkle hash tree (cached) semantics are identical.

This can be seen easily in, e.g., the Capella deposit operations tests:
```
nimbus-eth2/tests/consensus_spec/capella/test_fixture_operations.nim(88) test_fixture_operations
nimbus-eth2/tests/consensus_spec/capella/test_fixture_operations.nim(80) runTest
nimbus-eth2/vendor/nim-unittest2/unittest2.nim(900) testImpl
nimbus-eth2/tests/helpers/debug_state.nim(53) runTestX60gensym63
nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/assertions.nim(38) failedAssertImpl
nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/assertions.nim(28) raiseAssert
nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(54) sysFatal

    Unhandled defect: nimbus-eth2/tests/helpers/debug_state.nim(53, 15) ``==`(preState.inactivity_scores.hashes[idx`gensym106],
     postState.inactivity_scores.hashes[idx`gensym106])`
Diff: preState.inactivity_scores.hashes[0] = 00000000000000001e2809040101010101010201080101010101010101010101
and   postState.inactivity_scores.hashes[0] = 0000000000000000010709040101010101010201080101010101010101010101
 [AssertionDefect]
  [FAILED] [Valid]   EF - Capella - Operations - Deposit - effective_deposit_with_genesis_fork_version
```

https://github.com/status-im/nim-ssz-serialization/blob/cf2319bdff5c63d04b320f830dd7ea8b7e152bc6/ssz_serialization/types.nim#L292-L294
still holds, i.e. the current approach is correct.

If one really wanted to remove this, some of the available options:
- define the API to specify equivalence classes, that, well, even if these two `HashLists` don't have identical `hashes` members, they're functionally identical.
- define the API to simply require that the state can look inconsistent (with regard to the above) between a call from `clearCache`/`clearCaches` and the next `hash_tree_root` call.
- in a more ad hoc, way, modify the tests to be looser about this.
- factor out the in-band signaling to out-of-band signaling by storing this information elsewhere in the `HashList`. Most variations of this run into a similar sort of problem, though, where doing pure equality comparisons might reveal technically different state which is the same thing, just tracking the kind of node thing it is.
- Play even more subtle tricks, such as encoding this in the `indices`.

But I don't think these are worthwhile, at least unless/until this proves a bottleneck. it introduces too much technical risk to work around a Nim issue that I'd rather, if it becomes pressing, find a more direct workaround for, to express this initialization idiom more efficiently.

So I think it makes sense to leave this as is, for this PR.